scio-elasticsearch5/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,328 @@
+        checkNotNull(flushInterval);
+        input
+            .apply("Assign To Shard", ParDo.of(new AssignToShard<>(numOfShard)))
+            .apply("Re-Window to Global Window", Window.<KV<Long, T>>into(new GlobalWindows())
+                .triggering(Repeatedly.forever(
+                    AfterProcessingTime
+                        .pastFirstElementInPane()
+                        .plusDelayOf(flushInterval)))
+                .discardingFiredPanes())
+            .apply(GroupByKey.create())
+            .apply("Write to Elasticesarch",
+                ParDo.of(new ElasticsearchWriter<>
+                    (clusterName, servers, toDocWriteRequests, error)));
+        return PDone.in(input.getPipeline());
+      }
+    }
+
+    private static class AssignToShard<T> extends DoFn<T, KV<Long, T>> {
+
+      private final Long numOfShard;
+
+      public AssignToShard(Long numOfShard) {
+        this.numOfShard = numOfShard;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        // assign this element to a random shard
+        final long shard = ThreadLocalRandom.current().nextLong(numOfShard);
+        c.output(KV.of(shard, c.element()));
+      }
+    }
+
+    private static class ElasticsearchWriter<T> extends DoFn<KV<Long, Iterable<T>>, Void> {
+
+      private final Logger LOG = LoggerFactory.getLogger(ElasticsearchWriter.class);
+      private final ClientSupplier clientSupplier;
+      private final SerializableFunction<T, Iterable<DocWriteRequest<?>>> toDocWriteRequests;
+      private final ThrowingConsumer<BulkExecutionException> error;
+
+      public ElasticsearchWriter(String clusterName,
+                                 InetSocketAddress[] servers,
+                                 SerializableFunction<T, Iterable<DocWriteRequest<?>>> toDocWriteRequests,
+                                 ThrowingConsumer<BulkExecutionException> error) {
+        this.clientSupplier = new ClientSupplier(clusterName, servers);
+        this.toDocWriteRequests = toDocWriteRequests;
+        this.error = error;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        final List<Iterable<DocWriteRequest<?>>> docWriteRequests =
+            StreamSupport.stream(c.element().getValue().spliterator(), false)
+                .map(toDocWriteRequests::apply)
+                .collect(Collectors.toList());
+
+        // Elasticsearch throws ActionRequestValidationException if bulk request is empty,
+        // so do nothing if number of actions is zero.
+        if (docWriteRequests.isEmpty()) {
+          LOG.info("ElasticsearchWriter: no requests to send");
+          return;
+        }
+
+        final BulkRequest bulkRequest = new BulkRequest();
+        Iterables.concat(docWriteRequests)
+            .forEach(x -> {
+              try {
+                bulkRequest.add(x);
+              } catch(InvalidIndexNameException e) {
```

Comment:
`InvalidIndexNameException` already a RuntimeException there's no need for special treatment.

```
final BulkRequest bulkRequest = new BulkRequest().add(Iterables.concat(docWriteRequests));
```